packages/ethereum/contracts/src/Crypto.sol

```diff
@@ -0,0 +1,598 @@
+   * https://ethresear.ch/t/you-can-kinda-abuse-ecrecover-to-do-ecmul-in-secp256k1-today/2384
+   *
+   * @param scalar to multiply with secp256k1 base point
+   */
+  function scalarTimesBasepoint(uint256 scalar) public pure returns (address) {
+    return
+      ecrecover(
+        0,
+        SECP256K1_BASE_POINT_Y_COMPONENT_SIGN,
+        bytes32(SECP256K1_BASE_POINT_X_COMPONENT),
+        bytes32(mulmod(scalar, SECP256K1_BASE_POINT_X_COMPONENT, SECP256K1_FIELD_ORDER))
+      );
+  }
+
+  /**
+   * Computes multiplicative inverse of secp256k1 field element.
+   *
+   * Uses little Fermat because expmod precompile is more gas-efficient
+   * than extended euclidian algorithm in Solidity.
+   */
+  function invMod(uint256 el) public view returns (uint256 o) {
+    if (el == 0) {
+      revert InvalidFieldElement();
+    }
+
+    assembly {
+      //  define pointer
+      let p := mload(0x40)
+      // store data assembly-favouring ways
+      mstore(p, 0x20)             // Length of Base
+      mstore(add(p, 0x20), 0x20)  // Length of Exponent
+      mstore(add(p, 0x40), 0x20)  // Length of Modulus
+      mstore(add(p, 0x60), el)  // Base
+      mstore(add(p, 0x80), 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2D) // p - 2
+      mstore(add(p, 0xa0), 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F) // p
+      if iszero(staticcall(not(0), 0x05, p, 0xC0, p, 0x20)) {
+        revert(0, 0)
+      }
+
+      o := mload(p)
+    }
+  } 
+
+  /**
+   * Performs a modular exponentiation using the expmod precompile.
+   */
+  function expmod(uint256 base, uint256 exponent, uint256 modulus) public view returns (uint256 o) {
+    assembly {
+      //  define pointer
+      let p := mload(0x40)
+      // store data assembly-favouring ways
+      mstore(p, 0x20)             // Length of Base
+      mstore(add(p, 0x20), 0x20)  // Length of Exponent
+      mstore(add(p, 0x40), 0x20)  // Length of Modulus
+      mstore(add(p, 0x60), base)  // Base
+      mstore(add(p, 0x80), exponent)    // Exponent
+      mstore(add(p, 0xa0), modulus)     // Modulus
+      if iszero(staticcall(not(0), 0x05, p, 0xC0, p, 0x20)) {
+        revert(0, 0)
+      }
+
+      o := mload(p)
+    }
+  }
+
+  /**
+   * Converts a curve point to an Ethereum address.
+   *
+   * This function can be used to witness the result of a scalar
+   * multiplication.
+   */
+  function pointToAddress(CurvePoint calldata p) public pure returns (address) {
+    return address(uint160(uint256(keccak256(abi.encodePacked(p.x, p.y)))));
+  }
+
+  /**
+   * Adds two elliptic curve points using the general implementation.
+   *
+   * This function is optimized to perform one point addition.
+   *
+   * Optimizations:
+   * - solidity assembly
+   * - optimize for a single addition
+   */
+  function ecAdd(CurvePoint memory p, CurvePoint memory q) public view returns (CurvePoint memory r)  {
+    if (p.x == 0 && p.y == 0) { // Inf + Q = Q
```

Comment:
Identity is not (0,0) on secp256k1, it does not even satisfy the equation!
Same applies to the test below.

Reply:
changed